File: backend/src/models/Recipe.js
const mongoose = require('mongoose');

const recipeItemSchema = new mongoose.Schema({
  ingredientId: { type: mongoose.Schema.Types.ObjectId, ref: 'Ingredient', required: true },
  quantityRequired: { type: Number, required: true, min: 0 } // Amount deducted per product unit
});

const recipeSchema = new mongoose.Schema({
  productId: { type: mongoose.Schema.Types.ObjectId, ref: 'Product', required: true, unique: true },
  ingredients: [recipeItemSchema],
  instructions: { type: String, default: '' }
}, { timestamps: true });

module.exports = mongoose.model('Recipe', recipeSchema);

File: frontend/lib/src/models/recipe.dart
class Recipe {
  final String id;
  final String variantId;
  final String ingredientId;
  final double quantityRequired;

  Recipe({
    required this.id,
    required this.variantId,
    required this.ingredientId,
    required this.quantityRequired,
  });

  factory Recipe.fromMap(Map<String, Object?> map) {
    return Recipe(
      id: map['id'] as String,
      variantId: map['variant_id'] as String,
      ingredientId: map['ingredient_id'] as String,
      quantityRequired: (map['quantity_required'] as num).toDouble(),
    );
  }

  Map<String, Object?> toMap() {
    return {
      'id': id,
      'variant_id': variantId,
      'ingredient_id': ingredientId,
      'quantity_required': quantityRequired,
    };
  }
}
